daffodil-core/src/main/scala/org/apache/daffodil/grammar/GrammarTerm.scala

```diff
@@ -51,7 +52,8 @@ import org.apache.daffodil.dsom.Term
 abstract class Gram(contextArg: SchemaComponent)
   extends OOLAGHostImpl(contextArg)
   with BasicComponent
-  with GramRuntime1Mixin {
+  with GramRuntime1Mixin
+  with GramRuntime2Mixin {
```

Comment:
I think what we are looking for here is typeclasses. If the GramRuntime1Mixin was a typeclass, we could define it in the Runtime1 package; and provide an implementation in the Runtime1 package for all of the Gram objects while leaving the Gram objects themselves in the Core package.

Unfourtuantly, typeclasses are not a first class citizen in Scala; but they can be emulated using implicits. https://scalac.io/typeclasses-in-scala/

In the past, I have been dissapointed with this approach; but I believe that was largely a result of the implications with respect to boxing; which should not be an issue in this case.

An alternative approach would be to just turn parser() and generateCode() from methods of Gram (through a mixing) into simple functions in Runtime1/Runtime2 that take Gram as an arguement.

In either case, I would suggest that the way forward is that we implement Runtime2 within the Runtime2 package without adding a mixin to the Gram objects. Once that is done and we have a good feel for the method; we can discuss if it is worth refactoring Runtime1 as well.

Reply:
Well, by way of discussion..... bunch of points here.

So one issue is that runtime2 isn't independent of runtime1. I believe there are things created when runtime1 creates the parsers and unparsers that are used by runtime2. 

There's also a hypothesis here, that the backends could be pluggable because the schema compiler doesn't have to do backend-specific things. The schema compiler has a backend-independent aspect, which creates and hooks together the grammar objects. But the implementation of the grammar objects is the _essence_ of the back-end. 

A back end is (a) the part of the schema compiler (grammar objects) that create data structures for that back end, and (b) a library of primitives. 

Part (a) here is part of daffodil-core pretty much inherently right now. Grammar primitives use all the data structures of the schema compiler - the DSOM objects are available, and used as context for issuing SDEs from the grammar objects

Part  (b) is the daffodil-runtimeX module. For runtime2 right now this is a tiny thing currently, but would of course grow rapidly as we expand beyond just 32-bit integers. 

Every backend is going to add code to the grammar primitives. Right now those grammar primitives support 3 methods that backends implement. parser/unparser which create scala data structures that are invocable to carry out parsing/unparsing, and now generateCode, which carries out a generation processes, and like parser/unparser can cause SDEs to be issued, but otherwise is sort of the end of the line for the compiler.

If we wanted to separate backends in a pluggable way, from the rest of the schema compiler, I think the grammar objects would not have parser/unparser/code-generator methods on them at all. Instead the grammar data structure would be the output of the schema compiler. The back-end implementation of each grammar primitive's compilation would be found and delegated to, or maybe the grammar data structure would be walked/visited by a  separate back-end specific compiler pass.

The challenge with this refactoring is doing it without having to reimplement an echo of each grammar object class in each back-end.  

Object-oriented Mixins are tightly coupling everything together, but minimize this delegation and walker kind of "glue logic" which makes code so tedious to maintain. 

Given that runtime2 depends on runtime1 currently, this didn't seem worth it as of yet.

Reply:
This strikes me as a functional vs object-orientiented distinction.

In the functional view, the grammar objects are just a data structure. By definition the "implementation" is in functions which are defined seperately. It is then a purely organizational question if those functions should be defined near where the data type is defined.

I don't think there is much glue logic to be dealt with.

Setting aside the possibility of typeclasses, all we are talking about is a single parse() function with a match structure. This is a pattern we already use throughout the compiler. At least to me, the functions defined with this pattern are far easier to read, as all of the different versions of those functions can be found in one place; instead of needing to go through all of the different objects where different implementations of said functions are located.

Needing to maintain the "parser" function when its implementation is scattered throughout the code seems much more tedious than needing to update it whenever a grammar object is added/removed (which happens far less frequently than implementation changes).

With the addition of Runtime2; we are now putting parser logic, code generation logic, and abstract grammar logic all in the same spot. 

I agree that it may not be worth refactoring Runtime1. But we do not yet have any intertia behind Runtime2; so this is the best time to change how we implement it.

Reply:
I am coming around to brandon's way of thinking here. The various kinds of code aren't in the same spot file-wise, they're in mixins, and the mixins separate the concerns into different files. But they are hanging off the same classes, so in that sense, all in the same place. 

So for runtime2, we can try treating the grammar objects as just a data structure passed to the runtime2 back end "compiler", which then dispatches on type of grammar object. 

The class duplication I was worrying about is a false issue. The backend-specific functions don't have to mirror the grammar object constructors. The backend compiler functions  just get passed the grammar object and access things off of it.